### PR TITLE
add Tracer to skipper options (overrides OpenTracing)

### PR DESCRIPTION
**Pull request by gpradeepkrishna**
This PR enables applications that extends skipper to pre-initialize `Tracer` instance and feed it directly to Skipper overriding `OpenTracing`. The need for this arises due to the inherent behaviour of skipper making the `Tracer` available only via request context. Thus preventing the extending application to reuse the tracer instance.

### CODE REVIEW

File: skipper.go [hunk 1]
```diff
@@ -1468,20 +1472,13 @@ func run(o Options, sig chan os.Signal, idleConnsCH chan struct{}) error {
...
-		// existence of a tracer
-		tracer, _ = tracing.LoadTracingPlugin(o.PluginDirs, []string{"noop"})
+	tracer, err := initializeTracer(o)
+	if err != nil {
+		return err
 	}
 
 	// tee filters override if we have a tracer
-	if len(o.OpenTracing) > 0 {
+	if tracer != nil {
```
Comment (AlexanderYastrebov) [inline skipper.go:1481]: I think we can drop this condition since the above code ensures there is always a tracer available.

File: skipper.go [hunk 2]
```diff
@@ -2028,6 +2025,20 @@ func run(o Options, sig chan os.Signal, idleConnsCH chan struct{}) error {
 	return listenAndServeQuit(o.CustomHttpHandlerWrap(proxy), &o, sig, idleConnsCH, mtr, cr)
 }
 
+func initializeTracer(o Options) (ot.Tracer, error) {
```
Comment (AlexanderYastrebov) [inline skipper.go:2028]: Lets make it a method similar to `filterRegistry()`

File: skipper_test.go [hunk 3]
```diff
@@ -544,3 +544,32 @@ func TestDataClients(t *testing.T) {
 
 	sigs <- syscall.SIGTERM
 }
+
+func TestTracerInstanceTakesPrecedenceOverTracingOptions(t *testing.T) {
```
Comment (AlexanderYastrebov) [inline skipper_test.go:548]: Not sure if we need full integration test here, maybe lets make it similar to TestOptionsFilterRegistry

File: skipper.go (hunk 1 above)
Comment (gpradeepkrishna) [reply skipper.go:1481]: Agreed!

File: skipper.go (hunk 2 above)
Comment (gpradeepkrishna) [reply skipper.go:2028]: Makes sense, thanks

File: skipper_test.go (hunk 3 above)
Comment (gpradeepkrishna) [reply skipper_test.go:548]: Will amend this as well.

File: skipper_test.go [hunk 4]
```diff
@@ -109,6 +109,30 @@ func TestOptionsFilterRegistry(t *testing.T) {
...
+	tracer := &tracingtest.Tracer{}
+	// run skipper proxy that we want to test
+	o := Options{
+		Tracer:      tracer,
+		OpenTracing: []string{"noop"},
+	}
+
+	tr, err := o.tracerInstance()
+	assert.Nil(t, err)
+	assert.NotNil(t, tr)
```
Comment (AlexanderYastrebov) [inline skipper_test.go:122]:
```go
assert.NoError(t, err)
assert.Same(t, tracer, tr)
```
maybe also rename `tr` to `result` or `instance`

File: skipper_test.go [hunk 5]
```diff
@@ -109,6 +109,30 @@ func TestOptionsFilterRegistry(t *testing.T) {
...
+}
+
+func TestOptionsTracerInstanceFallbacksToOpenTracingWhenTracerIsNil(t *testing.T) {
+	// run skipper proxy that we want to test
+	o := Options{
+		OpenTracing: []string{"noop"},
+	}
+
+	tr, err := o.tracerInstance()
+	assert.Nil(t, err)
```
Comment (AlexanderYastrebov) [inline skipper_test.go:132]: assert.NoError(t, err)

File: skipper_test.go [hunk 6]
```diff
@@ -109,6 +109,28 @@ func TestOptionsFilterRegistry(t *testing.T) {
...
+		Tracer:      tracer,
+		OpenTracing: []string{"noop"},
+	}
+
+	tr, err := o.tracerInstance()
+	assert.Nil(t, err)
+	assert.Same(t, tr, tracer)
+}
+
+func TestOptionsTracerInstanceFallbacksToOpenTracingWhenTracerIsNil(t *testing.T) {
```
Comment (AlexanderYastrebov) [inline skipper_test.go:124]: Maybe just name it `TestOptionsOpenTracing`.

Lets also add a test (`TestOptionsDefaultTracer`) for the fallback tracer from empty `Options{}`

File: skipper_test.go [hunk 7]
```diff
@@ -109,6 +109,28 @@ func TestOptionsFilterRegistry(t *testing.T) {
...
+func TestOptionsTracerInstanceOverridesOpenTracing(t *testing.T) {
+	tracer := &tracingtest.Tracer{}
+	o := Options{
+		Tracer:      tracer,
+		OpenTracing: []string{"noop"},
+	}
+
+	tr, err := o.tracerInstance()
+	assert.NoError(t, err)
+	assert.Same(t, tr, tracer)
```
Comment (AlexanderYastrebov) [inline skipper_test.go:121]: expected shoudl be the second argument, i.e. `assert.Same(t, tracer, tr)` - that's why I suggested to rename one ore the other

File: skipper_test.go (hunk 6 above)
Comment (gpradeepkrishna) [reply skipper_test.go:124]: Looks like we overlooked a case where tracer can still be nil. This is when `o.Tracer = nil`, `o.OpenTracing = nil` & `pluginDirs = nil`.

Comment (AlexanderYastrebov) [reply skipper_test.go:124]: Hm, ok, then I guess we can return (nil, nil), have a test for it and restore nil check for tee filter override

Comment (gpradeepkrishna) [reply skipper_test.go:124]: alternatively we can also send back a noop tracer. WDYT?

File: skipper.go [hunk 8]
```diff
@@ -1145,6 +1149,24 @@ func (o *Options) tlsConfig(cr *certregistry.CertRegistry) (*tls.Config, error)
...
+	if o.Tracer != nil {
+		return o.Tracer, nil
+	}
+
+	if len(o.OpenTracing) > 0 {
+		return tracing.InitTracer(o.OpenTracing)
+	} else {
+		// always have a tracer available, so filter authors can rely on the
+		// existence of a tracer
+		tracer, _ := tracing.LoadTracingPlugin(o.PluginDirs, []string{"noop"})
```
Comment (AlexanderYastrebov) [inline skipper.go:1162]: The default option is to use `noop` so we do not need to create ot.NoopTracer{} so lets not swallow the error but simply `return tracing.LoadTracingPlugin(o.PluginDirs, []string{"noop"})`

File: skipper_test.go (hunk 6 above)
Comment (AlexanderYastrebov) [reply skipper_test.go:124]: I think we should change `LoadTracingPlugin` to return error if no plugin was loaded.

File: skipper.go (hunk 8 above)
Comment (AlexanderYastrebov) [reply skipper.go:1162]: Lets do this:
```go
if len(o.OpenTracing) > 0 {
  return tracing.InitTracer(o.OpenTracing)
} else {
  // always have a tracer available, so filter authors can rely on the
  // existence of a tracer
  tracer, err := tracing.LoadTracingPlugin(o.PluginDirs, []string{"noop"})
  if err != nil {
    return err
  } else if tracer == nil {
    // LoadTracingPlugin unfortunately may return nil tracer
    return fmt.Errorf("failed to load tracing plugin")
  }
  return tracer, nil
```

File: skipper_test.go (hunk 6 above)
Comment (AlexanderYastrebov) [reply skipper_test.go:124]: I propose https://github.com/zalando/skipper/pull/2525#discussion_r1299939623

**Comment by AlexanderYastrebov** [comment]
:+1:

File: skipper.go [hunk 9]
```diff
@@ -667,6 +667,10 @@ type Options struct {
 	// for a route when it's available (e.g. for RouteGroups)
 	OpenTracingBackendNameTag bool
 
+	// Tracer allows pre-created tracer to be passed on to skipper. Providing the tracer
+	// overrides options provided under OpenTracing property.
+	Tracer ot.Tracer
```
Comment (szuecs) [inline skipper.go:672]: @lucastt fyi, let's talk tomorrow about it

**Review by szuecs — COMMENTED** [review]
Let's talk tomorrow.
One question would be for a OT to OTel migration, how we would allow to set this. Should we have both OT and OTel tracer allowed to pass or should we only allow OTel and use the compatibility layer.

File: skipper.go (hunk 9 above)
Comment (lucastt) [reply skipper.go:672]: Same as the other comment, try to be specific with the naming `openTracingTracer` or something like that. Something that specifies that this is OpenTracing, not OpenTelemetry, so when we migrate to OpenTelemetry we support you getting one instance or the other.

File: skipper.go [hunk 10]
```diff
@@ -1145,6 +1149,27 @@ func (o *Options) tlsConfig(cr *certregistry.CertRegistry) (*tls.Config, error)
 	return config, nil
 }
 
+func (o *Options) tracerInstance() (ot.Tracer, error) {
```
[the remark this answers is not in the document]
Comment (gpradeepkrishna) [reply skipper.go]: Agreed and changed

**Comment by AlexanderYastrebov** [comment]
:+1:

**Comment by lucastt** [comment]
👍
